// mailosaur/esm/operations/devices.js
import DeviceListResult from '../models/deviceListResult.js';
import Device from '../models/device.js';
import OtpResult from '../models/otpResult.js';
class Devices {
    constructor(client) {
        this.client = client;
    }
    /**
     * Returns a list of your virtual security devices.
     */
    async list() {
        const url = `api/devices`;
        return new Promise((resolve, reject) => {
            this.client.request.get(url, {}, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new DeviceListResult(body));
            });
        });
    }
    /**
     * Creates a new virtual security device.
     * @param options Options used to create a new Mailosaur virtual security device.
     */
    async create(options) {
        const url = `api/devices`;
        return new Promise((resolve, reject) => {
            this.client.request.post(url, { body: options }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new Device(body));
            });
        });
    }
    /**
     * Retrieves the current one-time password for a saved device, or given base32-encoded shared secret.
     * @param query Either the unique identifier of the device, or a base32-encoded shared secret.
     */
    async otp(query) {
        if (!query || query.indexOf('-') > -1) {
            const url = `api/devices/${query}/otp`;
            return new Promise((resolve, reject) => {
                this.client.request.get(url, {}, (err, response, body) => {
                    if (err) {
                        return reject(err);
                    }
                    if (!response || response.statusCode !== 200) {
                        return reject(response
                            ? this.client.httpError(response)
                            : new Error('No response received'));
                    }
                    resolve(new OtpResult(body));
                });
            });
        }
        return new Promise((resolve, reject) => {
            const url = `api/devices/otp`;
            this.client.request.post(url, { body: { sharedSecret: query } }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new OtpResult(body));
            });
        });
    }
    /**
     * Permanently delete a virtual security device. This operation cannot be undone.
     * @param deviceId The unique identifier of the device.
     */
    async del(deviceId) {
        const url = `api/devices/${deviceId}`;
        return new Promise((resolve, reject) => {
            this.client.request.del(url, {}, (err, response) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 204) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve();
            });
        });
    }
}
export default Devices;
