// mailosaur/cjs/operations/devices.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const deviceListResult_js_1 = __importDefault(require("../models/deviceListResult.js"));
const device_js_1 = __importDefault(require("../models/device.js"));
const otpResult_js_1 = __importDefault(require("../models/otpResult.js"));
class Devices {
    constructor(client) {
        this.client = client;
    }
    /**
     * Returns a list of your virtual security devices.
     */
    async list() {
        const url = `api/devices`;
        return new Promise((resolve, reject) => {
            this.client.request.get(url, {}, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new deviceListResult_js_1.default(body));
            });
        });
    }
    /**
     * Creates a new virtual security device.
     * @param options Options used to create a new Mailosaur virtual security device.
     */
    async create(options) {
        const url = `api/devices`;
        return new Promise((resolve, reject) => {
            this.client.request.post(url, { body: options }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new device_js_1.default(body));
            });
        });
    }
    /**
     * Retrieves the current one-time password for a saved device, or given base32-encoded shared secret.
     * @param query Either the unique identifier of the device, or a base32-encoded shared secret.
     */
    async otp(query) {
        if (!query || query.indexOf('-') > -1) {
            const url = `api/devices/${query}/otp`;
            return new Promise((resolve, reject) => {
                this.client.request.get(url, {}, (err, response, body) => {
                    if (err) {
                        return reject(err);
                    }
                    if (!response || response.statusCode !== 200) {
                        return reject(response
                            ? this.client.httpError(response)
                            : new Error('No response received'));
                    }
                    resolve(new otpResult_js_1.default(body));
                });
            });
        }
        return new Promise((resolve, reject) => {
            const url = `api/devices/otp`;
            this.client.request.post(url, { body: { sharedSecret: query } }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new otpResult_js_1.default(body));
            });
        });
    }
    /**
     * Permanently delete a virtual security device. This operation cannot be undone.
     * @param deviceId The unique identifier of the device.
     */
    async del(deviceId) {
        const url = `api/devices/${deviceId}`;
        return new Promise((resolve, reject) => {
            this.client.request.del(url, {}, (err, response) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 204) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve();
            });
        });
    }
}
exports.default = Devices;
