// mailosaur/cjs/operations/messages.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mailosaurError_js_1 = __importDefault(require("../models/mailosaurError.js"));
const message_js_1 = __importDefault(require("../models/message.js"));
const messageListResult_js_1 = __importDefault(require("../models/messageListResult.js"));
const previewListResult_js_1 = __importDefault(require("../models/previewListResult.js"));
class Messages {
    constructor(client) {
        this.client = client;
    }
    /**
     * Waits for a message to be found. Returns as soon as a message matching the specified search criteria is found.
     * **Recommended:** This is the most efficient method of looking up a message, therefore we recommend using it wherever possible.
     * @param serverId The unique identifier of the containing server.
     * @param criteria The criteria with which to find messages during a search.
     * @param options Search options
     */
    async get(serverId, criteria, options = {}) {
        const getByIdError = new mailosaurError_js_1.default('Must provide a valid Server ID.', 'invalid_request');
        // Ensure we only return 1 result
        options.page = 0;
        options.itemsPerPage = 1;
        // Default timeout to 10s
        options.timeout = options.timeout || 10000;
        // Default receivedAfter to 1h
        options.receivedAfter =
            options.receivedAfter || new Date(Date.now() - 3600000);
        if (serverId.length !== 8) {
            return Promise.reject(getByIdError);
        }
        try {
            const result = await this.search(serverId, criteria, options);
            if (!result.items || result.items.length === 0) {
                throw new mailosaurError_js_1.default('No messages found matching criteria.', 'no_messages_found');
            }
            return await this.getById(result.items[0].id);
        }
        catch (err) {
            throw err;
        }
    }
    /**
     * Retrieves the detail for a single message. Must be used in conjunction with either list or
     * search in order to get the unique identifier for the required message.
     * @param messageId The unique identifier of the message to be retrieved.
     */
    async getById(messageId) {
        const url = `api/messages/${messageId}`;
        return new Promise((resolve, reject) => {
            this.client.request.get(url, {}, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new message_js_1.default(body));
            });
        });
    }
    /**
     * Permanently deletes a message. Also deletes any attachments related to the message. This operation cannot be undone.
     * @param messageId The identifier for the message.
     */
    async del(messageId) {
        const url = `api/messages/${messageId}`;
        return new Promise((resolve, reject) => {
            this.client.request.del(url, {}, (err, response) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 204) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve();
            });
        });
    }
    /**
     * Returns a list of your messages in summary form. The summaries are returned sorted by received date, with the most recently-received messages appearing first.
     * @param serverId The unique identifier of the required server.
     * @param options Message listing options
     */
    async list(serverId, options = {}) {
        const url = `api/messages`;
        const qs = {
            server: serverId,
            page: options.page,
            itemsPerPage: options.itemsPerPage,
            receivedAfter: options.receivedAfter,
            dir: options.dir,
        };
        return new Promise((resolve, reject) => {
            this.client.request.get(url, { qs }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new messageListResult_js_1.default(body));
            });
        });
    }
    /**
     * Permenantly delete all messages within a server.
     * @param serverId The unique identifier of the server.
     */
    async deleteAll(serverId) {
        const url = `api/messages`;
        const qs = {
            server: serverId,
        };
        return new Promise((resolve, reject) => {
            this.client.request.del(url, { qs }, (err, response) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 204) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve();
            });
        });
    }
    /**
     * Returns a list of messages matching the specified search criteria, in summary form.
     * The messages are returned sorted by received date, with the most recently-received messages appearing first.
     * @param serverId The unique identifier of the server to search.
     * @param criteria The criteria with which to find messages during a search.
     * @param options Search options
     */
    async search(serverId, criteria, options = {}) {
        const url = `api/messages/search`;
        let pollCount = 0;
        const startTime = Date.now();
        const qs = {
            server: serverId,
            page: options.page,
            itemsPerPage: options.itemsPerPage,
            receivedAfter: options.receivedAfter,
            dir: options.dir,
        };
        if (!Number.isInteger(options.timeout)) {
            options.timeout = 0;
        }
        if (typeof options.errorOnTimeout !== 'boolean') {
            options.errorOnTimeout = true;
        }
        const fn = (resolve, reject) => () => {
            this.client.request.post(url, { qs, body: criteria }, (err, response, body) => {
                var _a;
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                if (options.timeout &&
                    !body.items.length) {
                    const delayPattern = (((_a = response === null || response === void 0 ? void 0 : response.headers) === null || _a === void 0 ? void 0 : _a['x-ms-delay']) || '1000')
                        .split(',')
                        .map((x) => parseInt(x, 10));
                    const delay = pollCount >= delayPattern.length
                        ? delayPattern[delayPattern.length - 1]
                        : delayPattern[pollCount];
                    pollCount += 1;
                    // Stop if timeout will be exceeded
                    if (Date.now() - startTime + delay > options.timeout) {
                        return options.errorOnTimeout === false
                            ? resolve(new messageListResult_js_1.default(body))
                            : reject(new mailosaurError_js_1.default(`No matching messages found in time. By default, only messages received in the last hour are checked (use receivedAfter to override this). The search criteria used for this query was [${JSON.stringify(criteria)}] which timed out after ${options.timeout}ms`, 'search_timeout'));
                    }
                    return setTimeout(fn(resolve, reject), delay);
                }
                resolve(new messageListResult_js_1.default(body));
            });
        };
        return new Promise((resolve, reject) => {
            fn(resolve, reject)();
        });
    }
    /**
     * Creates a new message that can be sent to a verified email address. This is useful
     * in scenarios where you want an email to trigger a workflow in your product.
     * @param serverId The unique identifier of the required server.
     * @param options Options to use when creating a new message.
     */
    async create(serverId, options) {
        const url = `api/messages`;
        const qs = {
            server: serverId,
        };
        return new Promise((resolve, reject) => {
            this.client.request.post(url, { qs, body: options }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new message_js_1.default(body));
            });
        });
    }
    /**
     * @param messageId The unique identifier of the message to be forwarded.
     * @param options Options to use when forwarding a message.
     */
    async forward(messageId, options) {
        const url = `api/messages/${messageId}/forward`;
        return new Promise((resolve, reject) => {
            this.client.request.post(url, { body: options }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new message_js_1.default(body));
            });
        });
    }
    /**
     * Sends a reply to the specified message. This is useful for when simulating a user replying to one of your email or SMS messages.
     * @param messageId The unique identifier of the message to be forwarded.
     * @param options Options to use when replying to a message.
     */
    async reply(messageId, options) {
        const url = `api/messages/${messageId}/reply`;
        return new Promise((resolve, reject) => {
            this.client.request.post(url, { body: options }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new message_js_1.default(body));
            });
        });
    }
    /**
     * Generates screenshots of an email rendered in the specified email clients.
     * @param messageId The identifier of the email to preview.
     * @param options The options with which to generate previews.
     */
    async generatePreviews(messageId, options) {
        const url = `api/messages/${messageId}/screenshots`;
        return new Promise((resolve, reject) => {
            this.client.request.post(url, { body: options }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new previewListResult_js_1.default(body));
            });
        });
    }
}
exports.default = Messages;
