// mailosaur/cjs/operations/servers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const serverListResult_js_1 = __importDefault(require("../models/serverListResult.js"));
const server_js_1 = __importDefault(require("../models/server.js"));
class Servers {
    constructor(client) {
        this.client = client;
    }
    /**
     * Returns a list of your virtual servers. Servers are returned sorted in alphabetical order.
     */
    async list() {
        const url = `api/servers`;
        return new Promise((resolve, reject) => {
            this.client.request.get(url, {}, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new serverListResult_js_1.default(body));
            });
        });
    }
    /**
     * Creates a new virtual server.
     * @param options Options used to create a new Mailosaur server.
     */
    async create(options) {
        const url = `api/servers`;
        return new Promise((resolve, reject) => {
            this.client.request.post(url, { body: options }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new server_js_1.default(body));
            });
        });
    }
    /**
     * Retrieves the detail for a single server.
     * @param serverId The unique identifier of the server.
     */
    async get(serverId) {
        const url = `api/servers/${serverId}`;
        return new Promise((resolve, reject) => {
            this.client.request.get(url, {}, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new server_js_1.default(body));
            });
        });
    }
    /**
     * Retrieves the password for a server. This password can be used for SMTP, POP3, and IMAP connectivity.
     * @param serverId The unique identifier of the server.
     */
    async getPassword(serverId) {
        const url = `api/servers/${serverId}/password`;
        return new Promise((resolve, reject) => {
            this.client.request.get(url, {}, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(body.value);
            });
        });
    }
    /**
     * Updates the attributes of a server.
     * @param serverId The unique identifier of the server.
     * @param server The updated server.
     */
    async update(serverId, server) {
        const url = `api/servers/${serverId}`;
        return new Promise((resolve, reject) => {
            this.client.request.put(url, { body: server }, (err, response, body) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 200) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve(new server_js_1.default(body));
            });
        });
    }
    /**
     * Permanently delete a server. This will also delete all messages, associated attachments, etc. within the server. This operation cannot be undone.
     * @param serverId The unique identifier of the server.
     */
    async del(serverId) {
        const url = `api/servers/${serverId}`;
        return new Promise((resolve, reject) => {
            this.client.request.del(url, {}, (err, response) => {
                if (err) {
                    return reject(err);
                }
                if (!response || response.statusCode !== 204) {
                    return reject(response
                        ? this.client.httpError(response)
                        : new Error('No response received'));
                }
                resolve();
            });
        });
    }
    /**
     * Generates a random email address by appending a random string in front of the server's
     * domain name.
     * @param serverId The identifier of the server.
     */
    generateEmailAddress(serverId) {
        const host = process.env.MAILOSAUR_SMTP_HOST || 'mailosaur.net';
        const random = (Math.random() + 1).toString(36).substring(7);
        return `${random}@${serverId}.${host}`;
    }
}
exports.default = Servers;
